PlantUML Class Diagram

ElijahTheRandom opened the pull request.

Commit:
phantom file; test push.

ElijahTheRandom committed:
Create Class_Diagram

jonesz commented on doc/uml/Class_Diagram:
```diff
@@ -0,0 +1,49 @@
+@startuml
+class Data{
+-Idenitifer
+-Content
+}
```
Spelling should at least be fixed to "identifier". Looking below at "Tag", perhaps this should be called "name" to be consistent with a "rename()" method?

jonesz commented on doc/uml/Class_Diagram:
```diff
@@ -0,0 +1,49 @@
+-Relationship(Data, Tag)
++returnTag()
++returnData()
++changeRelationship()
```
Get rid of this. We can effectively change a relationship just through create/delete; the relationship doesn't do much other than connect data/tag so it's better viewed through an immutable light.

jonesz commented on doc/uml/Class_Diagram:
```diff
@@ -0,0 +1,49 @@
+}
+
+class Relationship{
+-Relationship(Data, Tag)
```
If this is a constructor, it should be public: `+Relatonship(data, tag)`. It might be better to assume constructors at this moment and just place fields with something akin to `data_ptr`, `tag_ptr`, which sort of explicitly tells you that this connects two objects that already exist.

jonesz commented on doc/uml/Class_Diagram:
```diff
@@ -0,0 +1,49 @@
+}
+
+class Constraint{
+~comparaters()
```
The PlantUML documentation indicates this is `package private` which I don't think is relevant. I get what you're trying to represent here I think, but I don't know how to do it myself; I just know this isn't it.

jonesz commented on doc/uml/Class_Diagram:
```diff
@@ -0,0 +1,49 @@
+Database --o Relationship
+Database --o Data
+Database --o Solver
+Database --o Tag
```
I think these go the other way around; `Database o-- Relationship`.

See 
![image](https://github.com/ElijahTheRandom/JavaProject321/assets/7352801/cbdc478d-ecf9-45b2-9063-365f2f722394)
The aggregation direction should be toward the `Database`. Should this be composition as well? I don't think we can have any of this stuff without the database: it's sort of a singleton that always exists.

jonesz commented on doc/uml/Class_Diagram:
```diff
@@ -0,0 +1,49 @@
+
+
+Solver --o Constraint
+@enduml
```
As a general comment, the UML spec says that fields and methods should begin with a lowercase letter and written in whateverCaseThisIsCalled.

jonesz commented on doc/uml/Class_Diagram:
```diff
@@ -0,0 +1,49 @@
+}
+
+class Tag{
+-Identifier
```
In general, the UML spec says that field names/methods should be written likeThisInAlternatingCase.

https://www.cs.uct.ac.za/mit_notes/software/htmls/ch05s09.html
> Names of attributes and operations start with a lower-case letter, but may have capital letters to indicate where the spaces would be, if the name would normally have spaces.

jonesz reviewed (changes requested):
This is a good start; like I said in Discord, there's probably multiple diagrams that can be made that show things at certain granularity (think a diagram with arrows with "connects" describing the relationship in plain English). We could showcase multiplicity on the specific associations (1-to-1, 1-to-N), etc. 

Since this will become more or less an artifact that goes in the appendix of the report, I think it's fair to just work on it incrementally and then at some point generate the images and attach them to the report.

:thumbs_up:

ElijahTheRandom committed:
Update Class_Diagram

Reworked to probably follow Ethan's comments

Commit:
Remove constructor, add fields.

At this point the constructor is likely implicit and the rest of
the UML classes follow a constructor-less design; utilize the
named fields instead.
